lib/internal/modules/ts/typescript.js

```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+const {
+  ArrayPrototypeSome,
+  String,
+  StringPrototypeEndsWith,
+} = primordials;
+const { transformSync } = require('internal/deps/swc/wasm');
+const {
+  stringify,
+} = require('internal/modules/helpers');
+const internalFsBinding = internalBinding('fs');
+const {
+  ERR_UNSUPPORTED_DIR_IMPORT,
+} = require('internal/errors').codes;
+const { fileURLToPath, URL } = require('internal/url');
+const { toNamespacedPath } = require('path');
+
+async function load(url, context, nextLoad) {
+  const isTS = context.format === 'typescript';
+  if (isTS) { context = { __proto__: context, format: 'module' }; }
+  const next = await nextLoad(url, context);
+  if (isTS) {
+    const { code } = transformSync(stringify(next.source), {
+      swcrc: false,
+      jsc: {
+        parser: {
+          syntax: 'typescript',
+        },
+        target: 'esnext',
+      },
+    });
+    return {
+      __proto__: null,
+      format: 'module',
+      source: code,
+    };
+  }
+  return next;
+}
+
+async function resolve(specifier, context, nextResolve) {
```

Comment:
nit: IMO it would be more readable to define `resolve` first, as it's the first one being called.